EKF/EKFGSF_yaw_estimator.cpp

```diff
@@ -0,0 +1,799 @@
+	Quatf quat_before_reset = _state.quat_nominal;
+	Quatf quat_after_reset = _state.quat_nominal;
+
+	// update transformation matrix from body to world frame using the current estimate
+	_R_to_earth = Dcmf(_state.quat_nominal);
+
+	// calculate the initial quaternion
+	// determine if a 321 or 312 Euler sequence is best
+	if (fabsf(_R_to_earth(2, 0)) < fabsf(_R_to_earth(2, 1))) {
+		// use a 321 sequence
+
+		Eulerf euler321(_state.quat_nominal);
+		euler321(2) = X_GSF[2];
+		quat_after_reset = Quatf(euler321);
+
+	} else {
+		// Calculate the 312 Tait-Bryan rotation sequence that rotates from earth to body frame
+		// We use a 312 sequence as an alternate when there is more pitch tilt than roll tilt
+		// to avoid gimbal lock
+		Vector3f rot312;
+		rot312(0) = X_GSF[2]; // first rotation about Z is taken from EKF-GSF
+		rot312(1) = asinf(_R_to_earth(2, 1)); // second rotation about X is taken from main EKF
+		rot312(2) = atan2f(-_R_to_earth(2, 0), _R_to_earth(2, 2));  // third rotation about Y is taken from main EKF
+
+		// Calculate the body to earth frame rotation matrix
+		const Dcmf R = taitBryan312ToRotMat(rot312);
+
+		// calculate initial quaternion states for the main EKF
+		// we don't change the output attitude to avoid jumps
+		quat_after_reset = Quatf(R);
+	}
+
+	// record the time for the magnetic field alignment event
+	_flt_mag_align_start_time = _imu_sample_delayed.time_us;
+
+	// calculate the amount that the quaternion has changed by
+	Quatf q_error =  quat_after_reset * quat_before_reset.inversed();
+	q_error.normalize();
+
+	// update quaternion states
+	_state.quat_nominal = quat_after_reset;
+	uncorrelateQuatStates();
+
+	// record the state change
+	_state_reset_status.quat_change = q_error;
+
+	// update transformation matrix from body to world frame using the current estimate
+	_R_to_earth = Dcmf(_state.quat_nominal);
+
+	// reset the rotation from the EV to EKF frame of reference if it is being used
+	if ((_params.fusion_mode & MASK_ROTATE_EV) && !_control_status.flags.ev_yaw) {
+		resetExtVisRotMat();
+	}
+
+	// update the yaw angle variance using half the nominal yaw separation between models
+	increaseQuatYawErrVariance(sq(fmaxf(M_PI_F / (float)N_MODELS_EKFGSF, 1.0e-2f)));
+
+	// add the reset amount to the output observer buffered data
+	for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
+		_output_buffer[i].quat_nominal = _state_reset_status.quat_change * _output_buffer[i].quat_nominal;
+	}
+
+	// apply the change in attitude quaternion to our newest quaternion estimate
+	// which was already taken out from the output buffer
+	_output_new.quat_nominal = _state_reset_status.quat_change * _output_new.quat_nominal;
+
+	// capture the reset event
+	_state_reset_status.quat_counter++;
+
+	// reset velocity and position states to GPS - if yaw is fixed then the filter should start to operate correctly
+	resetVelocity();
+	resetPosition();
+
+	ECL_INFO_TIMESTAMPED("EKF emergency yaw reset");
```

Comment:
The "EKF" in the beginning of the string is not really necessary, as the prefix "EKF" will be printed anyways. And it adds flash space. Same below.

Reply:
Will combine messages and fix.